Sign in with Hugging Face (OAuth support)

Wauplin opened the pull request:
This PR add OAuth support to Gradio a.k.a "Sign in with Hugging Face".

**Basic example:**
```py
import gradio as gr
from typing import Optional

def hello(profile: Optional[gr.OAuthProfile]) -> str:
    if profile is None:
        return "I don't know you."
    return f"Hello {profile.name}"

with gr.Blocks() as demo:
    gr.LoginButton()
    gr.LogoutButton()
    gr.Markdown().attach_load_event(hello, None)
```

**Space example:** https://huggingface.co/spaces/Wauplin/gradio-oauth-test

This is a demo to show how to require users to login and then check if they are on an authorize-list of users and display information only in that case.


## What's new? How it works?

First of all, `hf_oauth: true` must be set in the Space metadata. See docs for more details.

This PR adds two components, `gr.LoginButton` and `gr.LogoutButton`, and one object `gr.OAuthProfile`. When either the login or logout button is added to the Gradio app, 3 new routes are added to the FastAPI server:
- `/login/huggingface` that redirects the user to hf.co to login and grant access to the Space.
- `/login/callback` that handles the callback. At this point, user is identified and the server retrieve its profile.
- `/logout` that logs out the user by deleting its session cookie.

When a user triggers the login button, they are redirected to the HF login. Once back on the Gradio app, the button is disabled and the label get updated from `Sign in with Hugging Face` to `Signed in as Wauplin`. When the user is logged out, the login button switches back to its original state.

To access OAuth information, any Gradio function can accept `gr.OAuthProfile` or `Optional[gr.OAuthProfile]` as input. Gradio injects the user profile in the arguments the same way as `gr.Request` is handled:
- if `Optional[gr.OAuthProfile]`, the function is run no matter if the user is logged in or not. The logic of the function must handle both cases if user if logged in (profile available) or not (`None`)
- if `gr.OAuthProfile`, the function runs **only** if the user is logged in. This prevents unknown users to access parts of the app that are not public. If the user is not logged in, an exception is thrown _but I don't know how to nicely propagate the information to the front end._

## Limitations

1. When a user triggers the LoginButton, they are redirected to hf.co **in a new tab**. At the moment, running an OAuth app in the Space iframe can lead to errors in browsers that forbid cookies from third-party websites. Since the Space is hosted on `hf.co/spaces/...` but the embedded app is served on `....hf.space`, the cookies are not always forwarded (e.g. Safari users, but not only). If the user is already on `....hf.space`, the redirect is made in the same tab. For LogoutButton, it always happens in the same tab.
2. How to nicely display to the front-end that a function needs OAuth? Can be done in a another PR.
3. OAuth can only be used to retrieve user information but **not to perform actions of behalf of the user**. This will be implemented in the PR but is out of scope for this PR.
4. Would be awesome to customize the `gr.LoginButton` component to make it look [like this](https://huggingface.co/datasets/huggingface/badges/blob/main/sign-in-with-huggingface-xl-dark.svg). Any way of doing that ?
![image](https://github.com/gradio-app/gradio/assets/11801849/5b26d47e-5929-49c3-ad21-b02dc1ff3de8)

flozi00 commented:
If i can provide any help please tell me

The login module could prevent so much spam i get into my demos, looking forward to see that in production

Wauplin commented:
@abidlabs @freddyaboulton @aliabd @radames I've updated the implementation and completed the PR description. It is now in a state to be reviewed. I have tried different approaches on what to implement in gradio and what freedom to give to gradio users and I think this is a good balance. Gradio handles buttons to login/logout and injects user profiles in functions and the space maintainer can then decide what to do with it. In the end I did not implement a landing page to login as it wouldn't be flexible for the users + might need to evolve over time + would be complex to manage Spaces where OAuth is optional.

[Here is a demo Space](https://huggingface.co/spaces/Wauplin/gradio-oauth-test), please let me know what you think!

PS: I am still not sure what should be tested or not in Gradio, especially the full process with redirects and logins on HF. I'd appreciate suggestions as I'm not familiar with the test suite here :)

abidlabs commented on gradio/utils.py:
```diff
@@ -795,15 +796,19 @@ def get_type_hints(fn):
     if not hint:
         return False
     is_request = hint == Request
+    Optional[OAuthProfile]
+    # import pdb;pdb.set_trace()
```
```suggestion
```

abidlabs commented on requirements.txt:
```diff
@@ -1,12 +1,14 @@
 aiofiles>=22.0,<24.0
 aiohttp~=3.0
 altair>=4.2.0,<6.0
+authlib # required for oauth
```
Worried about dependency bloat. I wonder if we should maybe not add `authlib` and `itsdangerous` to the dependencies but instead raise an error if someone tries to use the OAuth-related features without them?

Wauplin replied:
I understand your concern, that's why I avoided to pin a dependency version. I was hoping it would be ok as they are quite shallow dependencies (`itsdangerous` is pure Python while `Authlib` only has a cryptography dep). 

```
Authlib==1.2.1
  - cryptography [required: >=3.2, installed: 38.0.1]
    - cffi [required: >=1.12, installed: 1.15.1]
      - pycparser [required: Any, installed: 2.21]
itsdangerous==2.1.2
```

Otherwise, would it be ok to define a group like  `gradio[oauth]` for example? And raise an issue to the user if it's not installed. Still not ideal IMO but better than asking to install individual packages.

Adding deps to Spaces would only be part of the solution IMO if we want to allow users to also debug things locally (related to [your comment](https://github.com/gradio-app/gradio/pull/4943#issuecomment-1660850609)).

abidlabs replied:
Yes, I like the `gradio[oauth]` + adding the dependencies to Spaces. Should catch most use cases without adding too many dependencies to `gradio`.

abidlabs commented on gradio/components/login_button.py:
```diff
@@ -0,0 +1,86 @@
+        self.attach_load_event(self._check_login_status, None)
+
+    def _check_login_status(self, request: Request) -> dict[str, Any]:
+        # Each time the page is refreshed or loaded, check if the user is logged in and adapt label
```
Nice!

abidlabs commented on gradio/components/login_button.py:
```diff
@@ -0,0 +1,86 @@
+        # ('self' value will be either "Sign in with Hugging Face" or "Signed in as ...")
+        self.click(fn=None, inputs=[self], outputs=None, _js=_js_open_if_not_logged_in)
+
+        # Attach load_event to check user login status when rendering
```
Clear from code imo
```suggestion
```

abidlabs commented on gradio/components/login_button.py:
```diff
@@ -0,0 +1,86 @@
+        elem_classes: list[str] | str | None = None,
+        scale: int | None = 0,
+        min_width: int | None = None,
+        _activate: bool = True,
```
Why do we have this parameter?

Wauplin replied:
I have genuinely no idea :smile: 
I took it from [`gr.DuplicateButton`](https://github.com/gradio-app/gradio/blob/f5539c7618e31451420bd3228754774da14dc65f/gradio/components/duplicate_button.py#L37) and thought it was an internal Gradio magic. Should I remove it?

abidlabs replied:
Yes its not really needed even in the case of `gr.DuplicateButton`, and is not needed here. I would just call `self.activate()` directly from the constructor.

abidlabs commented on gradio/helpers.py:
```diff
@@ -690,6 +690,26 @@ def special_args(
             if inputs is not None:
                 inputs.insert(i, request)
+        elif (
+            type_hint == Optional[oauth.OAuthProfile] or type_hint == oauth.OAuthProfile
```
Or `oauth.OAuthProfile | None`

Wauplin replied:
Handled in https://github.com/gradio-app/gradio/pull/4943/commits/103fba5ad947f14870b2a0934f5701c0625740eb

abidlabs commented on gradio/helpers.py:
```diff
@@ -690,6 +690,26 @@ def special_args(
+                    session["oauth_profile"] if "oauth_profile" in session else None
+                )
+                if type_hint == oauth.OAuthProfile and oauth_profile is None:
+                    raise ValueError(
```
You can raise a `gr.Error()` instead and it should propagate to the frontend

Wauplin replied:
Great! That's what I needed :)

Wauplin replied:
Added in https://github.com/gradio-app/gradio/pull/4943/commits/103fba5ad947f14870b2a0934f5701c0625740eb

abidlabs commented:
> Would be awesome to customize the gr.LoginButton component to make it look [like this](https://huggingface.co/datasets/huggingface/badges/blob/main/sign-in-with-huggingface-xl-dark.svg). Any way of doing that ?

Not at this time. We don't support icons in buttons, but it would be good to add support. Created an issue here: #5060

abidlabs commented:
Really great stuff @Wauplin! I left some comments above. 

One other point: right now, you can't run demos that include OAuth at all locally, which hampers the development experience of anyone trying this out. Would it be possible to either:

1. (Preferred) Provide users the values for env variables like `OAUTH_CLIENT_ID` so that they can set them locally and get the same experience as if running the Gradio app on Spaces
2. (If not...) at least allow the Gradio app to launch but have the OAuth part of the app non-functional -- the `gr.LoginButton` could simply raise a `gr.Error()` if the environment is not Spaces

Wauplin commented on requirements.txt:
```diff
@@ -1,12 +1,14 @@
 httpx
 huggingface_hub>=0.14.0
 importlib_resources>=1.3,<7.0
+itsdangerous # required for fastapi sessions
```
Yes it's a dependency for `SessionMiddleware`. Can't do without it .

Wauplin reviewed (commented):
Thanks for the review @abidlabs ! I will use `gr.Error` as suggested. For the deps issue, I replied in the comment. About enabling users to  debug their app locally, I agree it would be very nice to have a way of doing it. What I was thinking about is maybe to define dev secrets that would work only on localhost. I need to talk with the Hub team for that (I do think it's needed for OAuth dev in general, not only for Gradio). For now I tested things with a pair of secrets that I got internally but we should automatize that.

abidlabs commented:
Taking another look now @Wauplin

abidlabs commented on gradio/components/login_button.py:
```diff
@@ -0,0 +1,89 @@
+
+    def activate(self):
+        # Taken from https://cmgdo.com/external-link-in-gradio-button/
+        # Taking `self` as input to check if user is logged in
```
Clever, though maybe a bit fragile. Until we can support reading `State` variables in the frontend (#3525), this is fine.

abidlabs commented on gradio/components/logout_button.py:
```diff
@@ -0,0 +1,59 @@
+        elem_classes: list[str] | str | None = None,
+        scale: int | None = 0,
+        min_width: int | None = None,
+        _activate: bool = True,
```
Same -- would remove

abidlabs commented on gradio/components/logout_button.py:
```diff
@@ -0,0 +1,59 @@
+        _activate: bool = True,
+        **kwargs,
+    ):
+        super().__init__(
+            value,
+            variant=variant,
+            size=size,
+            icon=icon,
+            link=link,
+            visible=visible,
+            interactive=interactive,
+            elem_id=elem_id,
+            elem_classes=elem_classes,
+            scale=scale,
+            min_width=min_width,
+            **kwargs,
+        )
+        if _activate:
+            self.activate()
+
+    def activate(self):
+        # Redirect to logout page (which will delete the session cookie and redirect to
+        # landing page).
+        # Taken from https://cmgdo.com/external-link-in-gradio-button/
+        self.click(fn=None, _js="() => {window.location.assign('/logout');}")
```
You could use the `link` parameter in the `Button` class to make this simpler

abidlabs commented:
I was testing this on a Space to see what happens when you don't set the environmental variables, and it turns out that Spaces replaces the string `oauth` with `*****`, which is slightly confusing:

<img width="1359" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/07c646c3-8a06-4e27-b811-29b21658d8b9">

Would be good to find a way to correct this.

abidlabs commented:
I built a simple Space to test this out here: https://huggingface.co/spaces/abidlabs/oauth_test

I'm noticing two small issues after signing in (not sure if I'm using the buttons wrong):

1. The button text has not changed to reflect "Signed in as abidlabs" -- it remains "🤗 Sign in with HF"
2. After signing in, it reopens the original Space in a new tab but instead of opening the Space URL, it opens the URL of the underlying iframe. I think it would be better to stay in Spaces if possible

abidlabs commented:
One other thing -- I think its fine if the HF buttons don't work locally for now. But I think it would be better if an app with `gr.LoginButton`/`gr.LogoutButton` was still able to run locally so that someone could do local development (its hard to build a complicated gradio app directly on Spaces). We could just:

(1) Have clicking on the buttons not do anything
(2) Print a warning when the app runs letting the user know that the buttons will have no effect unless running on Spaces

abidlabs commented:
Finally, in terms of tests, I don't think we can do any E2E testing since this only works on Spaces. However, it might be worth writing a couple of unit tests to make sure events are attached correctly to the buttons. 

Similar to: https://github.com/gradio-app/gradio/blob/47294579294eb6db6ca2947cdcad9abffbdd7722/test/test_buttons.py#L4

Wauplin commented:
Thanks a lot for the review and suggestions @abidlabs! I worked on it today and here is the current state:

1. I removed `_activate` parameter and used `link` for the `LogoutButton` 
2. About dependencies, I have added a group `gradio[oauth]` to remove `authlib` and `itsdangerous` from the default dependencies. The problem now is that I don't know how to test it from the wheel package. We'll also have to update the Dockerfile template in Spaces. In the meantime, first users will have to add a `requirements.txt` themselves.
3. About [misleading error message](https://github.com/gradio-app/gradio/pull/4943#issuecomment-1670364431) with `hf_oauth` being hidden as `hf_*****`, I have created a PR to fix it server-side (https://github.com/huggingface/spaces-app-manager/pull/1033 - private repo). I don't think this is really blocking for this PR.
4. About [disabling buttons locally](https://github.com/gradio-app/gradio/pull/4943#issuecomment-1670382327), LoginButton and LogoutButton are now disabled and print a warning to the terminal when initialized locally.
5. I have added [a section in the documentation](https://github.com/gradio-app/gradio/pull/4943/files#diff-9ea01b5f76e63fe7804cf24cc5b47df546fdc353529cc32cb40e6b55a2f30ea9) with a snippet of code + a screenshot and linked a demo Space
6. I have added [a few tests](https://github.com/gradio-app/gradio/pull/4943/files#diff-29950609b4eb87049d6dd67329b5c4fcc04092d7d833ceb2f032fb6d14402dea) to check initialization and warnings.

Apart from that, [you said](https://github.com/gradio-app/gradio/pull/4943#issuecomment-1670373752) you've struggled to login on a test Space you've created. The thing is that the communication between the HF iframe and the Space itself is not perfect at the moment. If your browser blocks third-party libraries (e.g. Safari) it is currently NOT possible to _"Sign in with Hugging Face"_ while staying in the iframe tab. This is why we are always opening the sign in page on a second tab with a `...hf.space` URL. The short-term goal is to release this imperfect version of Gradio+OAuth and test it as it is (same as the [chat-ui Space](https://huggingface.co/spaces/coyotte508/chat-ui)). The mid-term goal is to definitely fix the cookie issue on Spaces either by working on the iframe <> window protocol (more complex) or by switching Spaces to URLs under `huggingface.co` (simpler but more side effects).
At the moment, those limitations are not documented. I have only documented the fact that this feature is experimental. As I see it, we should ship this first version and test it internally on real use cases while fixing the workflow issues (most probably after summer).

abidlabs commented on guides/01_getting-started/03_sharing-your-app.md:
```diff
@@ -145,7 +145,7 @@ This will add and document the endpoint `/api/addition/` to the automatically ge
 *Note*: For Gradio apps in which [queueing is enabled](https://gradio.app/guides/key-features#queuing), users can bypass the queue if they make a POST request to your API endpoint. To disable this behavior, set `api_open=False` in the `queue()` method. To disable the API page altogether, set `show_api=False` in `.launch()`.
 
-## Authentication
+## Password-protected app
```
Doing it this way just so the old anchor links (which we've shared in other places) continue to work

```suggestion
## Authentication

### Password-protected app
```

abidlabs commented on guides/01_getting-started/03_sharing-your-app.md:
```diff
@@ -166,6 +166,42 @@ demo.launch(auth=same_auth)
 For authentication to work properly, third party cookies must be enabled in your browser.
 This is not the case by default for Safari, Chrome Incognito Mode.
 
+## Enable OAuth
```
```suggestion
### Enable OAuth
```

abidlabs commented:
Changes look good @Wauplin and I like the plan of shipping an experimental version now while we work on making it a little more robust.

I tested out this demo locally:

```py
import gradio as gr

with gr.Blocks() as demo:
    gr.Markdown(
        "# Gradio OAuth Space\n\nThis Space is a demo for the new **Sign in with Hugging Face** feature.\n\nIt shows"
        " how you can use it to restrict access only to members of an organization. First it requires users to log in"
        " and then checks if the user meets the requirements (been part of"
        " [OAuthTesters](https://huggingface.co/OAuthTesters) in this case)\n\nSee"
        " https://github.com/gradio-app/gradio/pull/4943 for technical details."
    )
    with gr.Row():
        gr.LoginButton()
        gr.LogoutButton()

demo.launch()
```

and though the warnings are printed, I'm still getting a `ValueError` preventing me from seeing the app:

```
ValueError: OAuth is required but OAUTH_CLIENT_ID environment variable is not set. Make sure you've enabled OAuth in your Space by setting `hf_oauth: true` in the Space metadata.
```

abidlabs commented on gradio/components/login_button.py:
```diff
@@ -0,0 +1,96 @@
+    def activate(self):
+        if get_space() is None:
+            warnings.warn(
+                "At the moment, `LoginButton` is only available in a Gradio Space. When debugging an app locally, this button is disabled."
```
Where is the logic for disabling if not on Spaces?

abidlabs commented on guides/01_getting-started/03_sharing-your-app.md:
```diff
@@ -166,6 +166,42 @@ demo.launch(auth=same_auth)
+
+To enable OAuth, you must set `hf_oauth: true` as a Space metadata in your README.md file. This will register your Space
+as an OAuth application on Hugging Face. Next, you can use `gr.LoginButton` and `gr.LogoutButton` to add login and
+logout buttons to your Gradio app. Once a user is logged in with their HF account, you can retrieve their profile. To
```
They also need to include `itsdangerous` and `authlib` in a separate requirements.txt for now, right?

abidlabs replied:
Btw @Wauplin would you like to open PRs against the dockerfiles to add? https://github.com/huggingface/spaces-app-manager/tree/main/init_build/dockerfiles

Wauplin replied:
> They also need to include itsdangerous and authlib in a separate requirements.txt -- at least until we can include it in the Spaces dockerfile -- right?

Right! I added a line in the docs.

> Btw @Wauplin would you like to open PRs against the dockerfiles to add?

Yes I'm planning to do that but waiting for this PR to be merged and a version of gradio to be released first :+1: I'll take care of the PRs.

Wauplin replied:
Opened a PR: https://github.com/huggingface/spaces-app-manager/pull/1037

freddyaboulton commented:
> The problem now is that I don't know how to test it from the wheel package

Not sure if this is what you mean @Wauplin  but you can do `gradio[oauth] @ url` like so:

```bash
pip install "gradio[oauth] @ https://gradio-builds.s3.amazonaws.com/dc4086524d6402361dab047907f67f088b52a230/gradio-3.39.0-py3-none-any.whl"
```

Wauplin commented:
> you can do gradio[oauth] @ url like so

Thanks for the tip @freddyaboulton! That's very helpful to simplify the [requirements.txt](https://huggingface.co/spaces/Wauplin/oauth_test_from_abidlabs/blob/main/requirements.txt#L1) so far.

> I tried testing the latest version of this branch on Spaces, but now I'm seeing a different error: https://huggingface.co/spaces/abidlabs/oauth_test

@abidlabs Yep sorry for that the PR CI was failing to create the latest wheel package. CI is fixed so you can now install from latest release. I duplicated your Space and it works as expected for me (see [Space](https://huggingface.co/spaces/Wauplin/oauth_test_from_abidlabs) and [requirements.txt](https://huggingface.co/spaces/Wauplin/oauth_test_from_abidlabs/blob/main/requirements.txt)) 

> When testing this demo locally though the warnings are printed, I'm still getting a ValueError preventing me from seeing the app:

@abidlabs I changed today the behavior when running an app locally. Instead disabling the buttons, I have now mocked the login/logout workflow. If you run the Gradio demo locally and click on "Sign in with Hugging Face" you will get "logged in" with a fake user account. This way you can debug your demo locally without problem and once it's working, the deployment to a Space should happen seamlessly. A warning is still triggered locally to let the user know. I have also added a short paragraph to the documentation.

Could you now retry both locally and in your Space (https://huggingface.co/spaces/abidlabs/oauth_test) and check that your problems are solved. Please let me know if you see any missing part I should change.

freddyaboulton commented on gradio/components/login_button.py:
```diff
@@ -0,0 +1,88 @@
+
+
+@document()
+class LoginButton(Button):
```
Should we call these `HFLoginButton` and similarly for logout? I suspect there will be interest in this functionality for other auth providers and even the vanilla gradio auth

abidlabs replied:
I think we can add a parameter in `LoginButton` or `LogoutButton` specifying the authentication provider -- later once we have support for other auths.

abidlabs replied:
Although we do have the vanilla gradio auth at the moment, good point

abidlabs commented:
> @abidlabs I changed today the behavior when running an app locally. Instead disabling the buttons, I have now mocked the login/logout workflow. If you run the Gradio demo locally and click on "Sign in with Hugging Face" you will get "logged in" with a fake user account. This way you can debug your demo locally without problem and once it's working, the deployment to a Space should happen seamlessly. A warning is still triggered locally to let the user know. I have also added a short paragraph to the documentation.

Nice. I'll take a look right now

abidlabs commented:
Thanks @Wauplin tested locally and really like the mocked workflow.

I'm still not getting it to work successfully on Spaces. I tried your Space and although I'm able to click on the button and it asks me to authorize the app via HF, the button's text never changes to "Logged in as abidlabs". Here's a GIF to show what happens (I'm testing [your latest Space](https://huggingface.co/spaces/Wauplin/oauth_test_from_abidlabs)):

![Recording 2023-08-10 at 10 43 33](https://github.com/gradio-app/gradio/assets/1778297/a5079a96-1b28-4538-a81b-c04579c8351a)

abidlabs commented:
Update: I tested in incognito (thanks @Wauplin for the suggestion), and it's working now. So I presume there's something wonky about my browser settings or one of the extensions I've installed. 

Let me just look into why the functional tests are not passing.

abidlabs commented:
Did some further testing with @Wauplin and it looks like it was some previously saved cookies that were interfering with this. We could do things like delete previous cookies when a user tries to log in, but that could cause other issues, so we'll leave as is for now.

Wauplin commented:
Thanks a lot for all the help and reviews @abidlabs! I just pushed a commit to fix 1 of my tests but there are other failing tests indeed. Let me know if you find something.

abidlabs commented:
Yay all green ✅  Merging into `main` now @Wauplin
